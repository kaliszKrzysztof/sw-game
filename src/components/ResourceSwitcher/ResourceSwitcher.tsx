import React from 'react';
import { makeStyles } from '@material-ui/core/styles';
import Paper from '@material-ui/core/Paper';
import Tabs from '@material-ui/core/Tabs';
import Tab from '@material-ui/core/Tab';

type Item = {
  name: string;
  id: string;
  disabled: boolean;
};

interface ResourceSwitcherProps {
  items: Item[];
  activeItemId: string;
  onClick: (resource: string) => void;
}

const useStyles = makeStyles({
  root: {
    flexGrow: 1,
  },
});

const ResourceSwitcher: React.FC<ResourceSwitcherProps> = ({
  items,
  activeItemId,
  onClick,
}) => {
  const classes = useStyles();
  if (items.length === 0) {
    return null;
  }
  return (
    <Paper className={classes.root}>
      <Tabs
        value={activeItemId}
        onChange={(_, value) => onClick(value)}
        indicatorColor="primary"
        textColor="primary"
        centered
      >
        {items.map(({ name, id, disabled }) => (
          <Tab
            value={id}
            key={id}
            label={name}
            disabled={disabled}
            data-testid={id}
          />
        ))}
      </Tabs>
    </Paper>
  );
};

export default ResourceSwitcher;
