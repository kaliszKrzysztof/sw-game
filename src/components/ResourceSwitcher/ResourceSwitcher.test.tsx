import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import ResourceSwitcher from './ResourceSwitcher';

const items = [
  {
    name: 'Test1',
    id: 'test1',
    disabled: false,
  },
  {
    name: 'Test2',
    id: 'test2',
    disabled: false,
  },
];

describe('<ResourceSwitcher />', () => {
  it('should render', () => {
    render(
      <ResourceSwitcher
        items={items}
        activeItemId={items[0].id}
        onClick={() => null}
      />,
    );
  });
  it('should match snapshot', () => {
    const { container } = render(
      <ResourceSwitcher
        items={items}
        activeItemId={items[0].id}
        onClick={() => null}
      />,
    );
    expect(container).toMatchSnapshot();
  });

  it('can set initial tab', () => {
    const { getByTestId } = render(
      <ResourceSwitcher
        items={items}
        activeItemId={items[1].id}
        onClick={() => null}
      />,
    );
    expect(getByTestId(items[1].id)).toHaveAttribute('aria-selected', 'true');
  });

  it('can change active tab', () => {
    const defaultTab = items[0];
    const newTab = items[1];
    const { getByTestId, rerender } = render(
      <ResourceSwitcher
        items={items}
        activeItemId={defaultTab.id}
        onClick={() => null}
      />,
    );
    expect(getByTestId(defaultTab.id)).toHaveAttribute('aria-selected', 'true');
    rerender(
      <ResourceSwitcher
        items={items}
        activeItemId={newTab.id}
        onClick={() => null}
      />,
    );
    expect(getByTestId(newTab.id)).toHaveAttribute('aria-selected', 'true');
  });

  it('cannot click disabled items', () => {
    const mockOnClick = jest.fn();
    const itemsWithDisabledElement = [
      items[0],
      { ...items[1], disabled: true },
    ];
    const { getByTestId } = render(
      <ResourceSwitcher
        items={itemsWithDisabledElement}
        activeItemId={items[0].id}
        onClick={mockOnClick}
      />,
    );
    fireEvent.click(getByTestId(itemsWithDisabledElement[1].id));
    expect(mockOnClick).not.toHaveBeenCalled();
  });

  it('can click enabled items', () => {
    const mockOnClick = jest.fn();
    const { getByTestId } = render(
      <ResourceSwitcher
        items={items}
        activeItemId={items[1].id}
        onClick={mockOnClick}
      />,
    );
    fireEvent.click(getByTestId(items[0].id));
    expect(mockOnClick).toHaveBeenCalledWith(items[0].id);
  });
});
