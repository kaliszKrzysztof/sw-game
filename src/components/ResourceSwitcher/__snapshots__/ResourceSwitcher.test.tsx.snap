// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResourceSwitcher /> should match snapshot 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-root-6 MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      class="MuiTabs-root"
    >
      <div
        class="MuiTabs-scroller MuiTabs-fixed"
        style="overflow: hidden;"
      >
        <div
          class="MuiTabs-flexContainer MuiTabs-centered"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary Mui-selected"
            data-testid="test1"
            role="tab"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiTab-wrapper"
            >
              Test1
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
          <button
            aria-selected="false"
            class="MuiButtonBase-root MuiTab-root MuiTab-textColorPrimary"
            data-testid="test2"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <span
              class="MuiTab-wrapper"
            >
              Test2
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
        <span
          class="PrivateTabIndicator-root-7 PrivateTabIndicator-colorPrimary-8 MuiTabs-indicator"
          style="left: 0px; width: 0px;"
        />
      </div>
    </div>
  </div>
</div>
`;
